While in the process of cleaning data, we will often need to figure out whether we have duplicate data and if so we need to handle those and lets see how we can implement in pandas which is library in Python::

In [1]:
import pandas as pd

In [2]:
# read a dataset of movie reviewers by modifying the default parameter values like name of columns while read_table)
user_cols = ['user_id', 'age', 'gender', 'occupation', 'zip_code']
users = pd.read_table('http://bit.ly/movieusers', sep='|', header=None, names=user_cols, index_col='user_id')

In [3]:
users.head()

age gender  occupation zip_code
user_id                                 
1         24      M  technician    85711
2         53      F       other    94043
3         23      M      writer    32067
4         24      M  technician    43537
5         33      F       other    15213

This is a dataset of Movie Reviewers. 

In [4]:
users.shape

(943, 4)

shape says, this 'users' dataframe has 943 rows and 4 columns of data.

Suppose, we want to find duplicate 'zip_code's and will use Series method.

In [5]:
users.zip_code.duplicated().head(30)
#pandas.Series.duplicated : Duplicated values are indicated as True values in the resulting Series.

user_id
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29     True
30    False
Name: zip_code, dtype: bool

Here we found a Series of 'True's and 'False's and the logic here is, it returns 'True' like line 29 if there was an same record earlier to it - in other words there was identical record before line 29.

In [6]:
users.zip_code.head(30)

user_id
1     85711
2     94043
3     32067
4     43537
5     15213
6     98101
7     91344
8     05201
9     01002
10    90703
11    30329
12    06405
13    29206
14    55106
15    97301
16    10309
17    06355
18    37212
19    02138
20    95660
21    30068
22    40206
23    48197
24    94533
25    55107
26    21044
27    30030
28    55369
29    94043
30    55436
Name: zip_code, dtype: object

Here, when we try to check the value of Line 29 which is 94043 and try to find before Line 29, first value is found in Line 2 which is again 94043. 

Since this is a Series of boolean we can add them to count the number of duplicates by just saying sum().

In [7]:
users.zip_code.duplicated().sum()

148

Here, what happened is that 'True's will get converted to 1 and 'False's will get converted to 0 and 1's(True) adds it up and says 148 duplicate zip_code's.

###### Duplication in Dataframe: 

Duplication in Dataframe as a whole instead of just a Series. For instance :

In [8]:
users.duplicated()

user_id
1      False
2      False
3      False
4      False
5      False
       ...  
939    False
940    False
941    False
942    False
943    False
Length: 943, dtype: bool

What we will see here is that it will output a True if entire row is identical to the earlier row and as we saw in Series, here also we can count duplicate values using sum() method.

In [9]:
users.duplicated().sum()

7

There are 7 rows in which the values in that row are also the same values as a earlier row in the Dataframe. 

Now lets check those dupliacte rows how they look like by using Loc[] method and we passed a loc[] what row do we want to see, we can pass a Series of boolean and it will show any row in which True is present in that Series :

In [10]:
users.loc[users.duplicated(), :]

age gender occupation zip_code
user_id                                
496       21      F    student    55414
572       51      M   educator    20003
621       17      M    student    60402
684       28      M    student    55414
733       44      F      other    60630
805       27      F      other    20009
890       32      M    student    97301

Here, loc[ ] method allowed us to see 7 rows that were found by duplicate() method.

###### Lets get clarified about parameters in duplicated( ) method:: keep{‘first’, ‘last’, False}, default ‘first’

So if we say keep = 'first' as a string which is default (no change) but logic for 'first' is mark duplicates as True except for the first occurence. 

In other words the first occurence of record will remain 'False' (no duplicate) and all other identical record will change to 'True'(duplicate) in dataframe.

Method to handle dropping duplicates:

‘first’ : Mark duplicates as True except for the first occurrence.

‘last’ : Mark duplicates as True except for the last occurrence.

False : Mark all duplicates as True.

Here 'first' and 'last' are the Strings and False is boolean. 

In [11]:
users.loc[users.duplicated(keep='first'), :] # 'first' by default

age gender occupation zip_code
user_id                                
496       21      F    student    55414
572       51      M   educator    20003
621       17      M    student    60402
684       28      M    student    55414
733       44      F      other    60630
805       27      F      other    20009
890       32      M    student    97301

Resultset is same as before, it keep all the first occurrence records out of duplicate records.

In [12]:
users.loc[users.duplicated(keep='last'), :] 

age gender occupation zip_code
user_id                                
67        17      M    student    60402
85        51      M   educator    20003
198       21      F    student    55414
350       32      M    student    97301
428       28      M    student    55414
437       27      F      other    20009
460       44      F      other    60630

Resultset is different but number of rows are same and will keep all last occurrence records out of duplicate records.

In [13]:
users.loc[users.duplicated(keep=False), :] 

age gender occupation zip_code
user_id                                
67        17      M    student    60402
85        51      M   educator    20003
198       21      F    student    55414
350       32      M    student    97301
428       28      M    student    55414
437       27      F      other    20009
460       44      F      other    60630
496       21      F    student    55414
572       51      M   educator    20003
621       17      M    student    60402
684       28      M    student    55414
733       44      F      other    60630
805       27      F      other    20009
890       32      M    student    97301

" keep=False " : Mark all duplicates as True and its boolean and they will show up when we use Loc[ ] method.

Lets say, we would like to drop these duplicates from the Dataframe::

###### drop_duplicates( ) method::

In the above 'users' dataframe case, we dont actually believe that this user is same as the other user because they have unique user_id's but lets assume they are duplicates and we should drop from dataframe using drop_duplicates( ) method.

In [14]:
users.shape

(943, 4)

In [15]:
users.drop_duplicates(keep='first').shape

(936, 4)

Here, 'users' Dataframe lost 7 of duplicate rows using drop_duplicates() method but this wont occur 'inplace' by default. In other words dropping duplicate records wont be permanent without saying 'inplace=True' because by default it will be False.

And keep='first' means it will drop duplicates other than first occurred records.

keep='last' means it will drop duplicates other than last occurred records.

False : Will drop both versions('first' and 'last') of 14 rows of records(Drop all duplicates).

In [16]:
users.drop_duplicates(keep='last').shape

(936, 4)

In [17]:
users.drop_duplicates(keep=False).shape

(929, 4)

Here, it has dropped all 14 rows of duplicate records.

Lets commit the above changes by saying 'inplace = True'.

In [18]:
users.drop_duplicates(keep=False, inplace = True)

Now the changes took place in underlying dataframe too.

In [19]:
users.head()

age gender  occupation zip_code
user_id                                 
1         24      M  technician    85711
2         53      F       other    94043
3         23      M      writer    32067
4         24      M  technician    43537
5         33      F       other    15213

In [20]:
users.shape

(929, 4)

###### Useful tip:

What if we only wanted to consider certain columns when we are searching for duplicates.

Instead of all columns just need to look in for some columns.

Lets assume that " age + zip_code " should be unique identifier in this dataset.

In [21]:
users.duplicated(subset=['age', 'zip_code'])

user_id
1      False
2      False
3      False
4      False
5      False
       ...  
939    False
940    False
941    False
942    False
943    False
Length: 929, dtype: bool

Results Series of boolean.

In [22]:
users.duplicated(subset=['age','zip_code']).sum()

9

This means, if we say duplicated to only consider 'age' and 'zip_code' as relevant columns, there are 9 duplicates in the dataframe of this combination age and zip_code.

Similarly we can use drop_duplicates() with this subset as below:

In [23]:
users.drop_duplicates(subset=['age','zip_code']).shape

(920, 4)

Dataframe with those 9 rows that we found were dropped using drop_duplicates() method.